src/datepicker/datepicker-month-view.ts

```diff
@@ -1,51 +1,51 @@
 import {MonthViewModel, DayViewModel} from './datepicker-view-model';
-import {NgbDate} from './ngb-date';
+import {NgbMonthStruct} from './ngb-date-struct';
 import {NgbDatepickerI18n} from './datepicker-i18n';
-import {DayTemplateContext} from './datepicker-day-template-context';
+import {NgbDatepicker} from './datepicker';
 
 @Component({
   selector: 'ngb-datepicker-month-view',
   host: {'role': 'grid'},
   encapsulation: ViewEncapsulation.None,
   styleUrls: ['./datepicker-month-view.scss'],
   template: `
-    <div *ngIf="showWeekdays" class="ngb-dp-week ngb-dp-weekdays" role="row">
-      <div *ngIf="showWeekNumbers" class="ngb-dp-weekday ngb-dp-showweek"></div>
-      <div *ngFor="let w of month.weekdays" class="ngb-dp-weekday small" role="columnheader">
+    <div *ngIf="datepicker.showWeekdays" class="ngb-dp-week ngb-dp-weekdays" role="row">
+      <div *ngIf="datepicker.showWeekNumbers" class="ngb-dp-weekday ngb-dp-showweek"></div>
+      <div *ngFor="let w of viewModel.weekdays" class="ngb-dp-weekday small" role="columnheader">
         {{ i18n.getWeekdayShortName(w) }}
       </div>
     </div>
-    <ng-template ngFor let-week [ngForOf]="month.weeks">
+    <ng-template ngFor let-week [ngForOf]="viewModel.weeks">
       <div *ngIf="!week.collapsed" class="ngb-dp-week" role="row">
-        <div *ngIf="showWeekNumbers" class="ngb-dp-week-number small text-muted">{{ i18n.getWeekNumerals(week.number) }}</div>
+        <div *ngIf="datepicker.showWeekNumbers" class="ngb-dp-week-number small text-muted">{{ i18n.getWeekNumerals(week.number) }}</div>
         <div *ngFor="let day of week.days" (click)="doSelect(day)" class="ngb-dp-day" role="gridcell"
           [class.disabled]="day.context.disabled"
           [tabindex]="day.tabindex"
           [class.hidden]="day.hidden"
           [class.ngb-dp-today]="day.context.today"
           [attr.aria-label]="day.ariaLabel">
           <ng-template [ngIf]="!day.hidden">
-            <ng-template [ngTemplateOutlet]="dayTemplate" [ngTemplateOutletContext]="day.context"></ng-template>
+            <ng-template [ngTemplateOutlet]="datepicker.dayTemplate" [ngTemplateOutletContext]="day.context"></ng-template>
           </ng-template>
         </div>
       </div>
     </ng-template>
   `
 })
-export class NgbDatepickerMonthView {
-  @Input() dayTemplate: TemplateRef<DayTemplateContext>;
-  @Input() month: MonthViewModel;
-  @Input() showWeekdays;
-  @Input() showWeekNumbers;
+export class NgbDatepickerMonthView implements OnChanges {
+  @Input() month: NgbMonthStruct;
 
-  @Output() select = new EventEmitter<NgbDate>();
+  viewModel: MonthViewModel;
 
-  constructor(public i18n: NgbDatepickerI18n) {}
+  constructor(
+      public i18n: NgbDatepickerI18n, public datepicker: NgbDatepicker, private _elementRef: ElementRef<HTMLElement>) {}
+
+  ngOnChanges() { this.viewModel = this.datepicker.registerMonthView(this, this._elementRef.nativeElement); }
```

Comment:
Will it work if we inject the datepicker service here and remove `registerMonthView` from the datepicker completely?

So we'll just do these things right here inside the datepicker month view:
- handle focus events
- handle clicks
- inject a service